sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -2162,6 +2163,74 @@ object RemoveLiteralFromGroupExpressions extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * Prunes unnecessary fields from a [[Generate]] if it is under a count aggregation
+ * query.
+ */
+object CountAggregateOptimization extends Rule[LogicalPlan] {
```

Comment:
My understanding is, rule `ColumnPruning` will create `Project` with an empty project list under count-only aggregate. In the mean while, people can use DataFrame API to create empty `Project` as well.

I think it's simpler and more general to target `Project(Nil, Generate(...))`. We don't care about how the empty project was created, either by count-only aggregate, or by users.

Reply:
Do you mean just checking for empty project list?

Reply:
Oh I see. Got your point. You mean not just targeting count-only aggregation, but also generally empty project cases.

Reply:
> also generally empty project cases.

Yea, and not only empty, e.g. `Project(1, child)` can also trigger the optimization, as long as the Project has no references.